---
title: "TNBC_pipeline"
author: "Victoria Rios"
date: "14/6/2021"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data Analysis of TNBC gene expression information and mutations

## Data Pre-processing--------------------------------------------------------------------------------------------------

We avoid strings being recognized as factors
```{r}
options(stringsAsFactors = F)

library(factoextra)
library(M3C)
library(plyr)
```

### Expression data

Import ComBat resulting data frame containing gene expression information (previously normalized and batch effect assessed), where each column is a sample and each row a gene.
```{r}
# Change the working directory to the folder where the the file lives
setwd("C:\\Users\\victo\\OneDrive\\Documentos\\TNBC_research\\Database exploring")

# Read the csv file
train <- read.csv("ComBat/ComBat.csv", row.names = "row.names")

# TODO Delete after quick test
#test <- read.csv("ComBat/ComBat_test.csv", row.names = "row.names")
```

### Clinical information

Import ComBat compliant clinical information that matches with the samples found in the genetic information data frame.
```{r}
# Change the working directory to the folder where the the file lives
setwd("C:\\Users\\victo\\OneDrive\\Documentos\\TNBC_research\\Clinical_data_comparison")

# Read the csv file with the clinical information
combo_combat_mad <- read.csv("clinical_data_combat_mad.csv", row.names = "row.names")
RNA_clin_data <- combo_combat_mad
```

Make the categorical variables' values compliant

**Histological type---**
Checking the unique values of that clinical variable
```{r}
unique(RNA_clin_data$histological_type)
```

Mapping the values that need to be compliant
```{r}
RNA_clin_data$histological_type <- mapvalues(RNA_clin_data$histological_type, c("Mixed (not specified)"), c("Not specified"))

# Verify the changes have been made
unique(RNA_clin_data$histological_type)
```

**Histological grade---**
Checking the unique values of that clinical variable
```{r}
unique(RNA_clin_data$histological_grade)
```

Mapping the values that need to be compliant
```{r}
RNA_clin_data$histological_grade <- mapvalues(RNA_clin_data$histological_grade, c("UNKNOWN", "N/A"), c(NA, NA))

unique(RNA_clin_data$histological_grade)
```

**Race---**
Checking the unique values of that clinical variable
```{r}
unique(RNA_clin_data$race)
```

Mapping the values that need to be compliant
```{r}
RNA_clin_data$race <- mapvalues(RNA_clin_data$race, c("W", "B", "H"), c("not hispanic or latino", "not hispanic or latino", "hispanic or latino"))

unique(RNA_clin_data$race)
```

**Age-----**

Checking the unique values of that clinical value
```{r}
sort(unique(RNA_clin_data$age))
```

We assign each value to a bin according to the decade on their age and we save it as a column named "age_label"
```{r}
RNA_clin_data$age_label <- RNA_clin_data$age

for(age_row in 1:nrow(RNA_clin_data)){
  if(is.na(RNA_clin_data[age_row,"age"]) == FALSE){
    if(RNA_clin_data[age_row,"age"] < 40){
      RNA_clin_data[age_row,"age_label"] <- "30s" 
    }
    if(RNA_clin_data[age_row,"age"] >= 40 & RNA_clin_data[age_row,"age"] < 50){
      RNA_clin_data[age_row,"age_label"] <- "40s" 
    }
    if(RNA_clin_data[age_row,"age"] >= 50 & RNA_clin_data[age_row,"age"] < 60){
      RNA_clin_data[age_row,"age_label"] <- "50s" 
    }
    if(RNA_clin_data[age_row,"age"] >= 60 & RNA_clin_data[age_row,"age"] < 70){
      RNA_clin_data[age_row,"age_label"] <- "60s" 
    }
    if(RNA_clin_data[age_row,"age"] >= 70 & RNA_clin_data[age_row,"age"] < 80){
      RNA_clin_data[age_row,"age_label"] <- "70s" 
      }
    if(RNA_clin_data[age_row,"age"] >= 80 & RNA_clin_data[age_row,"age"] < 90){
      RNA_clin_data[age_row,"age_label"] <- "80s" 
    }
    if(RNA_clin_data[age_row,"age"] >= 90){
      RNA_clin_data[age_row,"age_label"] <- "90s" 
    }
  }
}

unique(RNA_clin_data$age_label)
```

**Age at menarche-----**
Checking the unique values of that clinical variable
```{r}
unique(RNA_clin_data$age_at_menarche)
```
The set of values are a small set, so we can skip the binning and use the values as they are.

**Tumor size-----**
Checking the unique values of that clinical variable
```{r}
sort(unique(RNA_clin_data$tumor_size_mm))
```

We assign each value to a bin according to the decade on their size and we save it as a column named "tumor_size"
```{r}
RNA_clin_data$tumor_size <- RNA_clin_data$tumor_size_mm

for(age_row in 1:nrow(RNA_clin_data)){
  if(is.na(RNA_clin_data[age_row,"tumor_size_mm"]) == FALSE){
    if(RNA_clin_data[age_row,"tumor_size_mm"] < 10){
      RNA_clin_data[age_row,"tumor_size"] <- "Below 10 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 10 & RNA_clin_data[age_row,"tumor_size_mm"] < 20){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 10 and 19 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 20 & RNA_clin_data[age_row,"tumor_size_mm"] < 30){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 20 and 29 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 30 & RNA_clin_data[age_row,"tumor_size_mm"] < 40){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 30 and 39 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 40 & RNA_clin_data[age_row,"tumor_size_mm"] < 50){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 40 and 49 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 50 & RNA_clin_data[age_row,"tumor_size_mm"] < 60){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 50 and 59 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 60 & RNA_clin_data[age_row,"tumor_size_mm"] < 70){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 60 and 69 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 70 & RNA_clin_data[age_row,"tumor_size_mm"] < 80){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 70 and 79 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 80 & RNA_clin_data[age_row,"tumor_size_mm"] < 90){
      RNA_clin_data[age_row,"tumor_size"] <- "Between 80 and 89 mm" 
    }
    if(RNA_clin_data[age_row,"tumor_size_mm"] >= 90){
      RNA_clin_data[age_row,"tumor_size"] <- "Greater than 89 mm" 
    }
  }
}

unique(RNA_clin_data$tumor_size)
```


## Dimensions reduction--------------------------------------------------------------------------------------------------

There is a great need to develop analytical methodology to analyze and to exploit the information contained in gene expression data. Because of the large number of genes and the complexity of biological networks, clustering is a useful exploratory technique for analysis of gene expression data. Prior starting the clustering process, principal component analysis (PCA) is needed to reduce the dimensionality of our large and complex data set and allow us to explain the data in only two dimensions.

**Principal components analysis (PCA)**
To compute a PCA in R we can use the prcomp() function. This function takes a matrix of data, where the columns are the variables that we want to use to transform our samples, which should be the rows of the matrix.

In our case, we want to look for similarities across our patients (samples = rows) based on gene expression (variables = columns). For that reason, we need to provide a transposed version of our table to the prcomp() function:
```{r}
t(train)[0:10,0:10]
```

```{r}
# Principal components analysis
x_train <- prcomp(t(train), scale = TRUE)

# TODO Delete after quick test
#x_train <- prcomp(t(cy_mat_combat[which(cy_mat_combat_mad > 1),]))
```

Visualize eigenvalues (scree plot) to show the percentage of variances explained by each principal component.
```{r}
scree_plot <- fviz_eig(x_train)
scree_plot

# TODO Delete after quick test
#fviz_eig(x_test)
```
As seen in the plot, we can explain most

Re-labeling the patients IDs
```{r}
# We check that the patient ID doesn't repeat
temp_train <- table(x_train[["x"]])
isTRUE(any(temp_train > 1)) # if the result is FALSE, they appear only once

# TODO remove
#temp_test <- table(x_test[["x"]])
#isTRUE(any(temp_test > 1)) # if the result is FALSE, they appear only once
```
We save the original IDs in another list inside of x
```{r}
orig_IDs_train <- matrix(row.names(x_train[["x"]]), ncol = 1) # we store the original IDs
short_IDs_train <- as.character(matrix(data = 1:nrow(x_train[["x"]]), ncol = 1)) # we store the short IDs
IDs_train <- cbind(orig_IDs_train, short_IDs_train) # we bind them in the same matrix
x_train[["IDs"]] <- IDs_train # We append the new IDs matrix to the x list

# TODO Remove
#orig_IDs_test <- matrix(row.names(x_test[["x"]]), ncol = 1) # we store the original IDs
#short_IDs_test <- as.character(matrix(data = 1:nrow(x_test[["x"]]), ncol = 1)) # we store the short IDs
#IDs_test <- cbind(orig_IDs_test, short_IDs_test) # we bind them in the same matrix
#x_test[["IDs"]] <- IDs_test # We append the new IDs matrix to the x list
```

We proceed to assign a unique number to each patient
```{r}
row.names(x_train[["x"]]) <- make.names(short_IDs_train, unique=TRUE)

head(x_train[["x"]])
```

**Graph of individuals**
We plot to visualize the short IDs of each patient 
```{r}
# train
fviz_pca_ind(x_train,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = T,     # Avoid text overlapping,
             max.overlaps = Inf
             )
```

Checking if there is a clear difference across the different components
```{r}
vals_to_check <- as.factor(RNA_clin_data$race)

plot(x_train$rotation[,1], x_train$rotation[,2],xlab="PCA component 1", ylab="PCA component 2", col=vals_to_check)
plot(x_train$rotation[,3], x_train$rotation[,4],xlab="PCA component 3", ylab="PCA component 4", col=vals_to_check)
plot(x_train$rotation[,5], x_train$rotation[,6],xlab="PCA component 5", ylab="PCA component 6", col=vals_to_check)
```
We can see that there isn't an apparent difference in the data explained by the first 6 components with the highest explained variance, so, we'll proceed with the analysis of the first two PCs.


## Consensus Clustering--------------------------------------------------------------------------------------------------------

*Extract taken from the M3C package documentation (https://rdrr.io/bioc/M3C/f/vignettes/M3Cvignette.Rmd)*
Genome-wide expression data is used to stratify patients into classes using clustering algorithms for precision medicine. The Monti consensus clustering algorithm (Monti et al., 2003) is a widely applied method to identify the number of clusters (K) through the principle of stability selection. This algorithm works by resampling and clustering the data for each K and a NXN consensus matrix is calculated, where each element represents the fraction of times two samples clustered together. A perfectly stable matrix would consist entirely of 0s and 1s, representing all sample pairs always clustering together or not together over resampling iterations. The next step is to compare the stability of these consensus matrices to decide K. The Proportion of Ambiguous Clustering (PAC) score (Senbabaoglu et al., 2014) has been proposed to assess consensus matrix stability for each K, however, it has bias towards greater values of K. This is due to a general problem with this type of consensus clustering algorithm that occurs because as K increases the consensus matrix converges towards a matrix of perfect stability simply by chance. The alternative well used delta K metric to find K is subjective as it relies on finding an elbow point and has been demonstrated to be inferior to the PAC score. Monte Carlo reference-based consensus clustering (M3C) (John et al., 2018) was made to solve these problems by comparing the real stability scores with those expected under a random model. M3C uses a Monte Carlo simulation to generate null distributions of stability scores along the range of K which, by comparing with the real stability scores, are used to decide the optimal K and reject the null hypothesis K=1.

```{r}
# Extract coordinates of individuals from the PCA result
ind.coord_train <- as.data.frame(get_pca_ind(x_train)$coord)

# TODO remove
#ind.coord_test <- as.data.frame(get_pca_ind(x_test)$coord)

head(ind.coord_train)
```

Performing the consensus clustering on the normalized data
```{r}
# We pass the gene expression data frame to perform consensus clustering
res_train <- M3C(train,
                 removeplots = TRUE,
                 iters = 10,
                 objective ='PAC',
                 fsize = 8,
                 lthick = 1,
                 dotsize = 1.25)

optimal_k_pac_train = max(res_train[["assignments"]]);

#TODO Remove
#res_test <- M3C(t(test), removeplots = TRUE, iters=3, objective='PAC', fsize=8, lthick=1, dotsize=1.25)
#optimal_k_pac_test = max(res_test[["assignments"]]);
```

(Optional) Save the result of consensus clustering for later
```{r}
# Save the result object to a local file
saveRDS(res_train,
        file = "consensus_clustering_res_final.rds")
# Restore the object
res_train <- readRDS(file = "consensus_clustering_res_final.rds")
```

### Checking the scores

The scores and p values are contained within the res$scores object. We can see below the RCSI reaches a maxima at K = 2, the p_score supports this optimal K decision. This means the null hypothesis that K = 1 can be rejected for this dataset because we have achieved significance (alpha=0.05) versus a dataset with no clusters.

```{r}
# train
res_train$scores
```

### Visualizing the consensus clustering plots

1. CDF Plot

In the CDF and following PAC plot we can see the inherant bias of consensus clustering where as K increases so does the apparent stability of the results (or CDF plot flatness), this we correct for by using a reference. This makes the method more sensitive to detection of the underlying structure in noisy data.

```{r}
res_train$plots[[1]]
```

2. PAC Score

This figure below shows the PAC score, we can see an elbow at K = 5 which is suggestive this is the best K. However, the bias of consensus clustering can be seen here as the PAC score naturally tends towards lower values as K increases (see above plot), making selecting K without taking this into account subject to bias. Selecting the minimal PAC score will only work when the clusters are very well seperated.

```{r}
res_train$plots[[2]]
```

3. Relative Cluster Stability Index (RCSI)

The Relative Cluster Stability Index (RCSI) was derived and associated 95% confidence intervals which take into account the reference PAC scores using the reference mean. This metric is better than the PAC score for deciding class number, where the maximum value corresponds to the optimal K. In this example the RCSI has an optima at K=2. Either the RCSI or p scores can be used to select K, in the Bioinformatics paper we used M3C with the p values to select K 

```{r}
res_train$plots[[4]]
```

4. P score

Finally, we calculate a p score from the distribution, here we display the p scores from the beta distribution. If none of the p scores reach significance over a reasonable range of K (e.g. 10), then we accept the null hypothesis. In the GBM dataset, we can see K = 2 reaches signficance with an alpha of 0.05, therefore we can reject the null hypothesis K=1 for the GBM dataset.
```{r}
res_train$plots[[3]]
```

# Entropy objective function

There is another method to find the ideal K value based on a different evaluation criteria (entropy, instead of the previously used 'PAC'). We use information entropy and aim to minimize it to find K, where lower values correspond to less uncertainty in the system (more stability during sampling). Whilst this function often gives pretty similar results to the PAC score, it makes better theoretical sense because it does not rely on a subjective window, neither does it rely on calculating a CDF first and a metric from this CDF second.
```{r}
res_entropy_train <- M3C(train,
                         method = 2,
                         fsize = 8,
                         lthick = 1,
                         dotsize = 1.25,
                         removeplots = TRUE,
                         objective = 'entropy')

optimal_k_entropy_train = max(res_entropy_train$assignments)

#res_entropy_test <- M3C(t(test), method = 2,fsize=8, lthick=1, dotsize=1.25, removeplots = TRUE)
#optimal_k_entropy_test = max(res_entropy_test$assignments)
```

Observing the results of the consensus clustering plots
```{r}
res_entropy_train$plots[[2]]
```

Comparing both consensus clustering methods optimal K
```{r}
if (optimal_k_pac_train == optimal_k_entropy_train){
  k_clusters_train = 2
  
  print(paste0("The optimal k clusters value is: ", as.character(k_clusters_train)))
}
```

Now we are pretty convinced there are 2 clusters within this data set which are not likely simply to have
occurred by chance alone. We can turn to examine the output objects that M3C generates.


# Understanding M3C Outputs

The 3 lines below extract the ordered (according in clustering results) expression data and the ordered annotation data from the results object after running M3C for a 2 cluster solution. We then take a look at the annotation object M3C outputs, a consensus cluster column has been added by M3C.

We gather the RNA-seq data and clinical data from the clustering results as different variables
```{r}
# RNA-seq data and cluster annotation
RNA_seq_data <- res_train$realdataresults[[2]][["ordered_data"]]
RNA_seq_cluster_annotation <- res_train$realdataresults[[2]][["ordered_annotation"]]

# Joining expression and cluster annotation
RNA_seq_merged <- as.data.frame(t(rbind(RNA_seq_data, t(RNA_seq_cluster_annotation))))

# Clinical data
RNA_clin_data <- merge(RNA_clin_data, RNA_seq_cluster_annotation, by="row.names")
rownames(RNA_clin_data) <- RNA_clin_data$Row.names
```

Save the results of the two clusters in a set of variables
```{r}
k_clusters_train = 2

data_train <- res_train$realdataresults[[k_clusters_train]]$ordered_data
annon_train <- res_train$realdataresults[[k_clusters_train]]$ordered_annotation    
ccmatrix_train <- res_train$realdataresults[[k_clusters_train]]$consensus_matrix

# TODO Remove
#data_test <- res_test$realdataresults[[k_clusters_test]]$ordered_data
#annon_test <- res_test$realdataresults[[k_clusters_test]]$ordered_annotation  
#ccmatrix_test <- res_test$realdataresults[[k_clusters_test]]$consensus_matrix
#head(annon)
```

Consensus matrix heatmaps from M3C output with ComplexHeatmap.
```{r}
 library(ComplexHeatmap)

 ccl <- list()
 ha_row <- list()
 x <- c("skyblue",
        "gold",
        "violet",
        "darkorchid",
        "slateblue",
        "forestgreen",
        "violetred",
        "orange",
        "midnightblue",
        "grey31",
        "black")
 
 names(x) <- as.character(seq(1,11,by=1))
 for (i in seq(2,k_clusters_train)){
   # get cc matrix and labels
   ccmatrix <- res_train$realdataresults[[i]]$consensus_matrix
   db_source <- as.data.frame(RNA_clin_data[c(rownames(res_train$realdataresults[[i]]$ordered_annotation)),"source_db"])
   rownames(db_source) <- c(rownames(res_train$realdataresults[[i]]$ordered_annotation))
   colnames(db_source) <- c("db_source")
   annon <- res_train$realdataresults[[i]]$ordered_annotation
   
   # do heatmap
   n <- 10
   seq <- rev(seq(0,255,
                  by=255/(n)))
   palRGB <- cbind(seq,
                   seq,255)
   mypal <- rgb(palRGB,
                maxColorValue=255)
   ha = HeatmapAnnotation(
   df= data.frame(Cluster=as.character(annon[,1])),
                  Source=as.character(db_source[,1]),
                  col = list(Cluster=x, 
                             Source = c("TCGA_RNAseq"="green",
                                        "Ref2_RNAseq"="red",
                                        "RefGSE20271_Array"="blue",
                                        "RefGSE575678_Array"="pink",
                                        "Ref4_RNAseq"="yellow")))
   print(head(ha))
   print(head(ccmatrix))
   ccl[[i]] <- Heatmap(ccmatrix,
                       name = "Consensus_index",
                       top_annotation = ha,
                       col=mypal,
                       show_row_dend = FALSE,
                       show_column_dend = FALSE,
                       cluster_rows = FALSE,
                       cluster_columns = FALSE,
                       show_column_names = FALSE)
 }
 
 print(ccl)
```
From looking at the heatmap, we can discard that the clustering was biased by the batch effect of the data sources.

**K-means clustering**
We proceed to apply a K-means clustering procedure on the PCA selected components (in this case, Dim 1 and Dim 2).
```{r}
k_train <- kmeans(ind.coord_train,
                  centers=2)
```

We plot the resulting clusters
```{r}
# Add clusters obtained using the K-means algorithm
ind.coord_train$cluster <- factor(k_train$cluster)
eigenvalue <- round(get_eigenvalue(x_train), 1)
variance.percent_train <- eigenvalue$variance.percent
```

```{r}
library(ggpubr)

ggscatter(
  ind.coord_train, 
  x = "Dim.1", 
  y = "Dim.2", 
  color = "cluster", 
  palette = "npg", 
  ellipse = TRUE, 
  ellipse.type = "convex", 
  size = 1.5,  
  legend = "right", 
  ggtheme = theme_bw(),
  xlab = paste0("Dim 1 (", variance.percent_train[1], "% )" ),
  ylab = paste0("Dim 2 (", variance.percent_train[2], "% )" )
) +
  stat_mean(aes(color = cluster), size = 4)
```

```{r}
library(ggrepel)
library(factoextra)
library(ggpubr)

histological_type <- as.factor(RNA_clin_data$histological_type)
ind.coord_train$histological_type <- histological_type

histological_grade <- as.factor(RNA_clin_data$histological_grade)
ind.coord_train$histological_grade <- histological_grade

source_db <- as.factor(RNA_clin_data$source_db)
ind.coord_train$source_db <- source_db


ggscatter(
  ind.coord_train, 
  x = "Dim.1", 
  y = "Dim.2", 
  palette = "npg", 
  ellipse = TRUE, 
  ellipse.type = "convex",
  size = 1.5,  
  legend = "right", 
  ggtheme = theme_bw(),
  color = "source_db",
  xlab = paste0("Dim 1 (", variance.percent_train[1], "% )" ),
  ylab = paste0("Dim 2 (", variance.percent_train[2], "% )" )
)
```
By looking at the previous plot, we confirm (once again), that the clustering was not biased due to the difference in data sources and gene expression techniques.

## Clinical data visualization per cluster

We split the clinical data in cluster 1 and cluster 2
```{r}
clin_data1 <- RNA_clin_data[which(RNA_clin_data["consensuscluster"] == 1),]
clin_data2 <- RNA_clin_data[which(RNA_clin_data["consensuscluster"] == 2),]
```

**Plotting categorical and numeric variables per cluster**
```{r}
library(glue)

for(vars in c("age_label", 
              "race", 
              "histological_grade", 
              "histological_type", 
              "tumor_size", 
              "response_to_treatment", 
              "pathological_history", 
              "family_pathological_hsitory", 
              "age", 
              "tumor_size_mm", 
              "age_at_menarche",
              "treatment",
              "vital_status",
              "chemotherapy")){
  var1 <- as.data.frame(table(clin_data1[vars]))
  cluster_info <- as.data.frame(rep("1",
                                    nrow(var1)))
  var1 <- cbind(cluster_info,
                var1)
  colnames(var1) <- c("cluster_info",
                      "variable", 
                      "count")
  
  
  var2 <- as.data.frame(table(clin_data2[vars]))
  cluster_info <- as.data.frame(rep("2",
                                    nrow(var2)))
  var2 <- cbind(cluster_info, var2)
  colnames(var2) <- c("cluster_info", 
                      "variable", 
                      "count")
  
  merged_df <- rbind(var1, 
                     var2)
  merged_df
  
  ## PLOTS
  library(ggplot2)
  
  # Grouped barcharts
  plt <- ggplot(merged_df, 
                aes(x=cluster_info, 
                    y=count, 
                    fill=variable)) + 
                geom_bar(position="dodge", 
                         stat="identity")
  
  if(vars %in% c("age", 
                 "tumor_size_mm", 
                "age_at_menarche")){
    merged_df <- rbind(clin_data1[c("consensuscluster", vars)], 
                       clin_data2[c("consensuscluster", vars)])
    colnames(merged_df) <- c("consensuscluster",
                             "vars")
    
    plt <- ggplot(merged_df, aes(x=vars, 
                                 group=consensuscluster, 
                                 fill=consensuscluster)) + 
      geom_density(alpha=.4) + ggtitle(vars)
  }
  
  print(plt)
}

```
Overall, at least visually, the plots seem to have a similar distribution of clinical values. The main highlights so far could be done on the proportions, e.g. 
-  the cluster no.2 presents:
   -  A drastically higher proportion of not hispanic or latino patients.
   -  A higher proportion of ductal histological type and is the only group among the two that presents the medular and metaplastic       histological type.
   -  Include a higher proportion of older patients, being the only cluster featuring patients of the 90 years old group.
   -  Response to treatment in this group seems to be equally poor or partial, meanwhile in the cluster 1 there are two instances of       progressive cases.
   -  A higher proportion of patients in this cluster are alive, compared to the proportions in cluster 1. 

## Significance tests on clinical information

We taje a look at the clinical data
```{r}
head(RNA_clin_data)
```

### Chi2 test on histological type

```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "histological_type"
values_to_check <- unique(RNA_clin_data$histological_type)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
    
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```


Doing the chi2 test on the entire variable (not per category only)
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "histological_type"
values_to_check <- unique(RNA_clin_data$histological_type)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$histological_type)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$histological_type)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), histological_type = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```

### Chi2 test on race

```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "race"
values_to_check <- unique(RNA_clin_data$race)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & 
                                  RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 &
                                  is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 &
                               is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 &
                               is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
    
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```

Doing the chi2 test on the entire feature
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "race"
values_to_check <- unique(RNA_clin_data$race)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$race)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$race)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), race = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```

### Chi2 test on histological grade
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "histological_grade"
values_to_check <- unique(RNA_clin_data$histological_grade)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```

Doing the chi2 test on the entire feature
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "histological_grade"
values_to_check <- unique(RNA_clin_data$histological_grade)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$histological_grade)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$histological_grade)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), histological_grade = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```

### Chi2 test on age
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "age_label"
values_to_check <- unique(RNA_clin_data$age_label)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
  print(value)
    # is the row cluster 1 or cluster 2
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
    
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```

Doing the chi2 test on the entire feature
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "age_label"
values_to_check <- unique(RNA_clin_data$age_label)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$age_label)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$age_label)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), age_label = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```


### Chi2 test on tumor stage
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "tumor_stage"
values_to_check <- unique(RNA_clin_data$tumor_stage)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
  print(value)
    # is the row cluster 1 or cluster 2
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
    
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
    # appending each name into a vector
    #mut_chisq_test_names <- c(mut_chisq_test_names, paste(rownames(RNA_clin_data)[mut]))
    #print(f_test_var$p.value)
  
    print(df)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    #rownames(mut_chisq_test) <- c("0", "1")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```

Doing the chi2 test on the entire feature
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "tumor_stage"
values_to_check <- unique(RNA_clin_data$tumor_stage)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$tumor_stage)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$tumor_stage)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), tumor_stage = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```

### Fisher test on age at menarche
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "age_at_menarche"
values_to_check <- unique(RNA_clin_data$age_at_menarche)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
  print(value)
    # is the row cluster 1 or cluster 2
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
    
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```

Doing the chi2 test on the entire feature
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "age_at_menarche"
values_to_check <- unique(RNA_clin_data$age_at_menarche)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$age_at_menarche)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$age_at_menarche)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), age_at_menarche = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```

## Fisher test on tumor size
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "tumor_size"
values_to_check <- unique(RNA_clin_data$tumor_size)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

for(value in c(values_to_check)){
  print(value)
    # is the row cluster 1 or cluster 2
    cluster1_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2_no <- length(which(RNA_clin_data[,variable_to_check] == value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster1 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 1 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    cluster2 <- length(which(RNA_clin_data[,variable_to_check] != value & RNA_clin_data[,"consensuscluster"] == 2 & is.na(RNA_clin_data[,variable_to_check]) != TRUE))
    
    df <- data.frame(cluster1=c(0,0), cluster2=c(0,0))
    
   
    df$cluster1 <- c(cluster1_no, cluster1)
    df$cluster2 <- c(cluster2_no, cluster2)
    rownames(df) <- c("0", "1")
    
    chisq_test_var <- chisq.test(df)
    mut_chisq_test <- c(chisq_test_var$p.value)
  
    print(df)
    
    # we save the P values and FDR of each test
    mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
    colnames(mut_chisq_test) <- c("p_value", "fdr")
    
    # we order the resulting matrix by P value
    print(value)
    print(mut_chisq_test)
}

```
Doing the chi2 test on the entire feature
```{r}
## Saving the possible outcomes of the feature
mut_chisq_test_names <- c()
variable_to_check <- "tumor_size"
values_to_check <- unique(RNA_clin_data$tumor_size)
values_to_check <- values_to_check[is.na(values_to_check) != TRUE]

# is the row cluster 1 or cluster 2
cluster1 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 1),]$tumor_size)
cluster2 <- table(RNA_clin_data[which(RNA_clin_data[,"consensuscluster"] == 2),]$tumor_size)

clusters_df <-  cbind(cluster1, cluster2)
clusters_df <- as.table(t(clusters_df))
dimnames(clusters_df) <- list(cluster = c("1", "2"), tumor_size = c(colnames(clusters_df)))

chisq_test_var <- chisq.test(clusters_df)
mut_chisq_test <- c(chisq_test_var$p.value)

# we save the P values and FDR of each test
mut_chisq_test <- cbind(mut_chisq_test, p.adjust(mut_chisq_test))
colnames(mut_chisq_test) <- c("p_value", "fdr")

# we order the resulting matrix by P value
print(chisq_test_var)
print(mut_chisq_test)

```



## Mutations Analysis--------------------------------------------------------------------------------------------------------
We download the mutations from the available sources

Ref 2 mutations
```{r}
mut_Ref2 <- read.delim(file = "C:\\Users\\victo\\OneDrive\\Documentos\\TNBC_research\\Database exploring\\DB compressed files\\Ref 2\\Ref2_untar\\brca_metabric\\data_mutations.txt", sep = "\t", header = T, stringsAsFactors = F, skip = 1)

head(mut_Ref2)
```

Ref 4 mutations
```{r}
mut_Ref4 <- read.delim(file = "C:\\Users\\victo\\OneDrive\\Documentos\\TNBC_research\\Database exploring\\DB compressed files\\Ref 4\\Ref4_untar\\brca_mbcproject_wagle_2017\\data_mutations.txt", sep = "\t", header = T, stringsAsFactors = F, skip = 0)

head(mut_Ref4)
```

TCGA mutations
```{r}
mut_TCGA <- read.delim(file = "C:/Users/victo/OneDrive/Documentos/R/TNBC research/TNBC_research/BRCA-TP.final_analysis_set.maf", sep = "\t", header = T, stringsAsFactors = F)

head(mut_TCGA)
```

Merging the mutation datasets
```{r}
# Columns in common with all the mutation datasets
merged_mut_cols <- intersect(colnames(mut_Ref2), colnames(mut_Ref4))
merged_mut_cols <- intersect(merged_mut_cols, colnames(mut_TCGA))

# Merged mutation datasets
merged_mut <- rbind(mut_Ref2[merged_mut_cols], mut_Ref4[merged_mut_cols])
merged_mut <- rbind(merged_mut, mut_TCGA[merged_mut_cols])

head(merged_mut)
```

We check the data frame dimensions
```{r}
dim(merged_mut)
```

We drop mutations related to: UTR, introns, RNA and silent.
```{r}
mut_filters <- c("3'UTR", "5'Flank", "3'Flank","5'UTR", "IGR", "Intron", "Silent", "RNA")
mut_filter_index <- which(merged_mut$Variant_Classification %in% mut_filters)
merged_mut <- merged_mut[-mut_filter_index,]

dim(merged_mut)
```

We can observe the type of mutation in the column "Variant_Classification"
```{r}
table(merged_mut$Variant_Classification)
```
We obtain the different samples and genes to make an array where it will store 0 and 1 depending on whether
mutation is present.
```{r}
u_samples <- sort(unique(merged_mut$Tumor_Sample_Barcode))
u_genes <- sort(unique(merged_mut$Hugo_Symbol))
mut_matrix <- matrix(0, nrow=length(u_genes), ncol=length(u_samples),

dimnames = list(u_genes, u_samples))

mut_matrix[1:5, 1:5]

mut_matrix <- mut_matrix[,-1]

mut_matrix[1:5, 1:5]
```

We fill the array by doing one loop per sample
```{r}
for(i in colnames(mut_matrix)){
aux_genes <- merged_mut$Hugo_Symbol[which(merged_mut$Tumor_Sample_Barcode == i)]
mut_matrix[aux_genes, i] <- 1
}
```


Defining the cluster groups with the patient IDs
```{r}
library(dplyr)
k_clusters = 2
cluster = list()

for (i in seq(1, k_clusters)) { 
                                cluster[[i]] <- filter(annon, annon$consensuscluster == i)
                                print(cluster[i]) 
                              }
```

We observe the difference in the naming convention of mutations and rna-seq data, as well as the difference in dimensions. We can compare the first 16 characters.
```{r}
raw_brca_rnaseq <- RNA_seq_merged

mut_rnaseq_match <- match(gsub(pattern = "-", replacement = "\\.",
x = substr(colnames(mut_matrix), 1, 16)),
substr(rownames(raw_brca_rnaseq), 1, 16))

mut_rnaseq_match
```

Clean the mutation matrix and expression matrix from NA values
```{r}
raw_brca_rnaseq <- t(RNA_seq_merged)

# to store the non NA values postitions
mut_rnaseq_match_nonNA <- which(!is.na(mut_rnaseq_match))

# to remove the NA values from mut_matrix_rna and mut_rnaseq_match
mut_matrix_rna <- mut_matrix[,mut_rnaseq_match_nonNA]
mut_rnaseq_match <- mut_rnaseq_match[mut_rnaseq_match_nonNA]

rna_mut <- raw_brca_rnaseq[,mut_rnaseq_match]

cbind(colnames(mut_matrix_rna)[c(1,20,50,54)], colnames(rna_mut)[c(1,20,50,54)])
```

We filter the genes in RNA-seq that show no variation and we keep the most frequent mutations only.
```{r}
# We convert all the dataframe as numeric
rna_mut <- mutate_all(data.frame(rna_mut), function(x) as.numeric(as.character(x)))

# We apply the mad function
rna_mut_mad <- apply(rna_mut, 1, mad)
mad_zero <- which(rna_mut_mad == 0)
rna_mut <- rna_mut[-mad_zero,]
table(apply(mut_matrix_rna,1,sum))
```

We count the times each mutation appears in the matrix and store the ones with an appearance of >9 times
```{r}
# We count the times each mutation appears in the matrix
mut_matrix_rna_count <- apply(mut_matrix_rna,1,sum)

# if the mutation count is greater than 9, we add it to the mut_matrix_rna
mut_matrix_rna <- mut_matrix_rna[which(mut_matrix_rna_count > 9),]

mut_matrix_rna
```

## T-test mutations merged

We run a t-test to know which genes change according to the mutations
```{r}
mut_rna_test <- c()
mut_rna_test_names <- c()
for(i in 1:nrow(mut_matrix_rna)){
# we apply the function "function(x)" to each of the row values of rna_mut where each row value is x
aux_mut_test <- t(sapply(1:nrow(rna_mut), function(x) {

# mutation: the mutation row (presence or absence of mutation per gene sequence)
# exp: the gene row of the rna_mut matrix
aux_test <- t.test(formula=exp~mut, data=data.frame(mut=mut_matrix_rna[i,],
exp=as.matrix(rna_mut)[x,]));

aux_test$p.value
}))

# appending each test into a vector
mut_rna_test <- c(mut_rna_test, aux_mut_test)
# appending each name into a vector
mut_rna_test_names <- c(mut_rna_test_names, paste(rownames(mut_matrix_rna)[i], rownames(rna_mut), sep="_"))

}

# we save the P values and FDR of each test
mut_rna_test <- cbind(mut_rna_test, p.adjust(mut_rna_test))
colnames(mut_rna_test) <- c("p_value", "fdr")
rownames(mut_rna_test) <- mut_rna_test_names

# we order the resulting matrix by P value
mut_rna_test <- mut_rna_test[order(mut_rna_test[,"p_value"]),]
mut_rna_test[1:5,]

```

Export the data frame with the p-values to a csv file
```{r}
mut_rna_test <- as.data.frame(mut_rna_test)
mut_rna_test$row_names <- rownames(mut_rna_test)

write.csv(as.data.frame(mut_rna_test),
          "C:\\Users\\victo\\OneDrive\\Documentos\\TNBC_research\\Clustering\\Mutations_merged_T_test.csv", 
          row.names = FALSE)
```

## T-test mutations per cluster
```{r}
mut_matrix_rna <- as.data.frame(mut_matrix_rna)
head(mut_matrix_rna)
```

```{r}
head(rna_mut)
```

```{r}
RNA_seq_merged <- as.data.frame(t(RNA_seq_merged))
head(RNA_seq_merged)
```

```{r}
colnames(RNA_seq_merged[,which(RNA_seq_merged["consensuscluster",] == 2)])
```

```{r}
colnames(RNA_seq_merged[,which(RNA_seq_merged["consensuscluster",] == 1)])
```

```{r}
mut_rnaseq_match_cluster1 <- match(gsub(pattern = "-", replacement = "\\.",
                          x = substr(colnames(rna_mut), 1, 16)),
                          substr(colnames(RNA_seq_merged[,which(RNA_seq_merged["consensuscluster",] == 1)]), 1, 16))

# to store the NA values postitions
NA_cluster1 <- which(is.na(mut_rnaseq_match_cluster1))

# to remove the NA values from mut_matrix_rna and mut_rnaseq_match
mut_matrix_rna_cluster1 <- mut_matrix_rna[,-NA_cluster1]
mut_rnaseq_match_cluster1 <- mut_rnaseq_match_cluster1[-NA_cluster1]
rna_mut_cluster1 <- rna_mut[-NA_cluster1]

print(dim(rna_mut_cluster1))
print(dim(mut_matrix_rna_cluster1))
```
```{r}
mut_rnaseq_match_cluster2 <- match(gsub(pattern = "-", replacement = "\\.",
                          x = substr(colnames(rna_mut), 1, 16)),
                          substr(colnames(RNA_seq_merged[,which(RNA_seq_merged["consensuscluster",] == 2)]), 1, 16))

# to store the NA values postitions
NA_cluster2 <- which(is.na(mut_rnaseq_match_cluster2))

# to remove the NA values from mut_matrix_rna and mut_rnaseq_match
mut_matrix_rna_cluster2 <- mut_matrix_rna[,-NA_cluster2]
mut_rnaseq_match_cluster2 <- mut_rnaseq_match_cluster2[-NA_cluster2]
rna_mut_cluster2 <- rna_mut[-NA_cluster2]

print(dim(rna_mut_cluster2))
print(dim(mut_matrix_rna_cluster2))
```


CLUSTER 1 
---------
We run a t-test to know which genes change according to the mutations 
```{r}
mut_rna_test <- c()
mut_rna_test_names <- c()
for(i in 1:nrow(mut_matrix_rna_cluster1)){
#for(i in 1:2){
# we apply the function "function(x)" to each of the row values of rna_mut where each row value is x
aux_mut_test <- t(sapply(1:nrow(rna_mut_cluster1), function(x) {

# mutation: the mutation row (presence or absence of mutation per gene sequence)
# exp: the gene row of the rna_mut matrix
aux_test <- t.test(formula=exp~mut, data=data.frame(mut=as.matrix(mut_matrix_rna_cluster1)[i,],
exp=as.matrix(rna_mut_cluster1)[x,]));

aux_test$p.value
}))

# appending each test into a vector
mut_rna_test <- c(mut_rna_test, aux_mut_test)
# appending each name into a vector
mut_rna_test_names <- c(mut_rna_test_names, paste(rownames(mut_matrix_rna_cluster1)[i], rownames(rna_mut_cluster1), sep="_"))

}

# we save the P values and FDR of each test
mut_rna_test <- cbind(mut_rna_test, p.adjust(mut_rna_test))
colnames(mut_rna_test) <- c("p_value", "fdr")
rownames(mut_rna_test) <- mut_rna_test_names

# we order the resulting matrix by P value
mut_rna_test <- mut_rna_test[order(mut_rna_test[,"p_value"]),]
mut_rna_test[1:5,]

```

CLUSTER 2 
---------
We run a t-test to know which genes change according to the mutations 
```{r}
mut_rna_test <- c()
mut_rna_test_names <- c()
for(i in 1:nrow(mut_matrix_rna_cluster2)){
#for(i in 1:2){
# we apply the function "function(x)" to each of the row values of rna_mut where each row value is x
aux_mut_test <- t(sapply(1:nrow(rna_mut_cluster2), function(x) {

# mutation: the mutation row (presence or absence of mutation per gene sequence)
# exp: the gene row of the rna_mut matrix
  
aux_test <- t.test(formula=exp,mut, data=data.frame(mut=as.matrix(mut_matrix_rna_cluster2)[i,],
exp=as.matrix(rna_mut_cluster2)[x,]));

aux_test$p.value
}))

# appending each test into a vector
mut_rna_test <- c(mut_rna_test, aux_mut_test)
# appending each name into a vector
mut_rna_test_names <- c(mut_rna_test_names, paste(rownames(mut_matrix_rna_cluster2)[i], rownames(rna_mut_cluster2), sep="_"))

}

# we save the P values and FDR of each test
mut_rna_test <- cbind(mut_rna_test, p.adjust(mut_rna_test))
colnames(mut_rna_test) <- c("p_value", "fdr")
rownames(mut_rna_test) <- mut_rna_test_names

# we order the resulting matrix by P value
mut_rna_test <- mut_rna_test[order(mut_rna_test[,"p_value"]),]
mut_rna_test[1:5,]

```

## Chi2 mutations merged

We run a chi2 to know which genes change according to the mutations
```{r}
mut_rna_test <- c()
mut_rna_test_names <- c()
for(i in 1:nrow(mut_matrix_rna)){
  # mutation: the mutation row (presence or absence of mutation per gene sequence)
  # exp: the gene row of the rna_mut matrix
  ones1 <- sum(mut_matrix_rna_cluster1[i,])
  zeros1 <- length(mut_matrix_rna_cluster1[i,])-ones1
  
  ones2 <- sum(mut_matrix_rna_cluster2[i,])
  zeros2 <- length(mut_matrix_rna_cluster2[i,])-ones2
  
  mut_clust <- as.table(rbind(c(zeros1, zeros2), c(ones1, ones2)))
  dimnames(mut_clust) <- list(status = c("No mutation","Mutation"),
                              cluster = c("1","2"))
  
  aux_mut_test <- chisq.test(mut_clust)$p.value
  
  # appending each test into a vector
  mut_rna_test <- c(mut_rna_test, aux_mut_test)
  # appending each name into a vector
  mut_rna_test_names <- c(mut_rna_test_names, paste(rownames(mut_matrix_rna)[i]))
}

# we save the P values and FDR of each test
mut_rna_test <- cbind(mut_rna_test, p.adjust(mut_rna_test))
colnames(mut_rna_test) <- c("p_value", "fdr")
rownames(mut_rna_test) <- mut_rna_test_names

# we order the resulting matrix by P value
mut_rna_test <- mut_rna_test[order(mut_rna_test[,"p_value"]),]
mut_rna_test[1:15,]
```